From: Jan-Marek Glogowski <jan-marek.glogowski@muenchen.de>
Date: Tue, 18 May 2010 17:47:05 +0200
Subject: Switch to lt_dlopenadvise() so back_perl can be opened with RTLD_GLOBAL.    
 Open all modules with RTLD_GLOBAL, needed so that back_perl can load
 non-trivial Perl extensions that require symbols from back_perl.so itself.
Bug-Debian: http://bugs.debian.org/327585

---
Index: openldap/servers/slapd/module.c
===================================================================
--- openldap.orig/servers/slapd/module.c	2022-05-20 17:36:16.057249110 -0400
+++ openldap/servers/slapd/module.c	2022-05-20 17:36:16.053249107 -0400
@@ -117,6 +117,20 @@
 	return -1;	/* not found */
 }
 
+static lt_dlhandle slapd_lt_dlopenext_global( const char *filename )
+{
+	lt_dlhandle handle = 0;
+	lt_dladvise advise;
+
+	if (!lt_dladvise_init (&advise) && !lt_dladvise_ext (&advise)
+			&& !lt_dladvise_global (&advise))
+		handle = lt_dlopenadvise (filename, advise);
+
+	lt_dladvise_destroy (&advise);
+
+	return handle;
+}
+
 int module_load(const char* file_name, int argc, char *argv[])
 {
 	module_loaded_t *module;
@@ -179,7 +193,7 @@
 	 * to calling Debug. This is because Debug is a macro that expands
 	 * into multiple function calls.
 	 */
-	if ((module->lib = lt_dlopenext(file)) == NULL) {
+	if ((module->lib = slapd_lt_dlopenext_global(file)) == NULL) {
 		error = lt_dlerror();
 #ifdef HAVE_EBCDIC
 		strcpy( ebuf, error );
